feat(pagination): add new pages templating options

ExFlo opened the pull request:
Before submitting a pull request, please make sure you have at least performed the following:

 - [ ] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) and [DEVELOPER.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/DEVELOPER.md) guide.
 - [ ] built and tested the changes locally.
 - [ ] added/updated any applicable tests.
 - [ ] added/updated any applicable API documentation.
 - [ ] added/updated any applicable demos.

ExFlo committed:
feat(pagination): add new pages templating options

ExFlo commented on src/pagination/pagination.spec.ts:
```diff
@@ -196,7 +320,7 @@ describe('ngb-pagination', () => {
       expectPages(fixture.nativeElement, ['«', '1', '+2', '3', '»']);
 
-      getLink(fixture.nativeElement, 0).click();
+      getLink(fixture.nativeElement, 1).click();
```
Hi here I fix the test that was not testing what we wanted (click on the page number) as previously it was clicking on the previous link (another test is doing it).
BR

maxokorokov commented on demo/src/app/components/pagination/demos/customization/pagination-customization.html:
```diff
@@ -6,4 +6,35 @@
+                class="form-control custom-pages-input"
+                id="paginationInput"
+                [value]="page"
+                (keyup.enter)="applyPage(myInputPage)"
```
can't we pass `(page: string)` and not `(myInputPage: HTMLInputElement)` here for simplicity?

so it would be something like `#inputPage (keyUp.enter) ="applyPage(inputPage.value)"`

maxokorokov commented on demo/src/app/components/pagination/demos/customization/pagination-customization.ts:
```diff
@@ -10,4 +12,21 @@ export class NgbdPaginationCustomization {
+    if (myInputPage.value === '') {
+      value = 1;
+    } else {
+      value = parseInt(myInputPage.value, 10);
```
what about just `parseInt() || 1` ? pages start with 1, not 0 anyway

maxokorokov commented on src/pagination/pagination.spec.ts:
```diff
@@ -145,6 +150,45 @@ describe('ngb-pagination', () => {
         () => { TestBed.configureTestingModule({declarations: [TestComponent], imports: [NgbPaginationModule]}); });
 
+    it('should render and respond to collectionSize change with customPages', () => {
+      const html = `<ngb-pagination [collectionSize]="collectionSize" [page]="1" [pageSize]="pageSize">
```
1. a **minimal** test case would be better... ex. I don't think we case about aria attibutes, classes, etc
2. instead of repeating the same template for 4 tests I think I would have extracted a `describe` block for this with a common test component

maxokorokov commented on src/pagination/pagination.spec.ts:
```diff
@@ -781,6 +949,21 @@ class TestComponent {
 
   onPageChange = () => {};
+  // User can apply the logic they want.
+  applyPage(myInputPage: HTMLInputElement) {
```
1. I dont think there are tests that actually use these `applyPage` and `formatInput` functions
2. as said before, I would have extracted another test component for this as it is pretty custom

maxokorokov commented on src/pagination/pagination.ts:
```diff
@@ -111,6 +112,16 @@ export class NgbPaginationPrevious {
+ */
+@Directive({selector: 'ng-template[ngbPaginationPages]'})
+export class NgbPaginationPages {
+  constructor(public templateRef: TemplateRef<NgbPaginationLinkContext>) {}
```
I think it merits a separate context like `NgbPaginationPagesContext` even though it might be a copy-paste of the link one at the moment.

maxokorokov reviewed (changes requested):
Hey, @ExFlo!

Sorry for the delay!

I think we can merge this, but I have a couple of comments:

Code:
- [ ] I think we should introduce `NgbPaginationPagesContext` for this feature as well

Tests:
- [ ] A refactoring would be nice → let's have a minimal common test component for the 4 cases you're testing
- [ ] Please add at least one test that changes the page from inside the custom page template

Demo:
- [ ] Added a couple of suggestions
- [ ] It doesn't look nice at the moment, can we make the input smaller ?
![Screenshot 2021-03-10 at 10 44 48](https://user-images.githubusercontent.com/8074436/110609581-a72c4400-818d-11eb-9d1b-fca1581f54e1.jpg)

Cheers,
Max

maxokorokov committed:
refactor(pagination): clean up `NgbPaginationPages` implementation

- clean up the `NgbPaginationPagesContext` and docs
- add a test for the `disabled` property in the context
- simplify demo even more

maxokorokov commented:
@ExFlo, I've added [a commit](https://github.com/ng-bootstrap/ng-bootstrap/pull/3910/commits/309cdfd13bf91bc6eb0616a03e365b5ee55650d3) here that:
- fixes formatting (your build was failing)
- fixes documentation comments
- fixes and removes inheritance between `Pages` and `Link` contexts (they're very different)
- simplifies demo even more
- fixes the styling for the demo (spacing, input size):

![Screenshot 2021-03-15 at 15 46 35](https://user-images.githubusercontent.com/8074436/111171962-b2c0a600-85a5-11eb-9530-c8c92346844a.jpg)
